### PR TITLE
image sign using per user registries.d

### PR DESCRIPTION
**Pull request by QiWang19**
Support per user  ~/.config/containers/registries.d to allow rootless image sign configurations.

Signed-off-by: Qi Wang <qiwan@redhat.com>

### CODE REVIEW

**Comment by rhatdan** [comment]
@QiWang19 Rebase and you should be able to pass the tests.

File: docs/source/markdown/podman-image-sign.1.md
Comment (rhatdan) [inline docs/source/markdown/podman-image-sign.1.md:13]: Where will signatures be written for non root users?

$HOME/.local/share/containers/sigstore?

Comment (QiWang19) [reply docs/source/markdown/podman-image-sign.1.md:13]: non-root users should create their YAML files under the $HOME/.config/containers/registries.d. Do we provide the default config file like default.yaml?

Comment (rhatdan) [reply docs/source/markdown/podman-image-sign.1.md:13]: Non root users are not going to know how to do this, so we need to setup good defaults.  podman should automatically detect that there is no default.yaml, and create it and store signatures in $HOME/.local/share/containers/sigstore.  If default.yaml exists, then it will follow those rules.  I would guess.

Comment (QiWang19) [reply docs/source/markdown/podman-image-sign.1.md]: Set the default sigstore to `$HOME/.local/share/containers/sigstore` if the config file is using `/var/lib/containers/sigsore`. Also hard code this path if rootless user has no configurations at all.

**Comment by QiWang19** [comment]
@mtrmac @rhatdan PTAL

**Comment by rhatdan** [comment]
LGTM
@mtrmac PTAL

**Comment by rhatdan** [comment]
@mtrmac @vrothberg PTAL

File: docs/source/markdown/podman-image-sign.1.md
Comment (vrothberg) [inline docs/source/markdown/podman-image-sign.1.md:43]:
```suggestion
configuration files in /etc/containers/registries.d/ for root, or $HOME/.config/containers/registries.d for non-root users.  When you sign
```

File: docs/source/markdown/podman-image-sign.1.md
Comment (vrothberg) [inline docs/source/markdown/podman-image-sign.1.md:14]: The comment doesn't seem to match the code. We don't seem to check for permissions but use the rootless paths if they exist even for root.

**Review by vrothberg — COMMENTED** [review]
I really want a head nod from @mtrmac. I may very well miss details/side effects I am may not be aware of.

File: pkg/trust/trust.go
Comment (mtrmac) [inline pkg/trust/trust.go:88]: Should these two be a top-level commented variables like the others?

File: pkg/domain/infra/abi/images.go
Comment (mtrmac) [inline pkg/domain/infra/abi/images.go:739]: Dealing with URLs here seems to be an unnecessary complication; move the `localPathFromURI` call to the `registryInfo != nil` case, and this can just work with the `sigStoreDir` path.

**Review by mtrmac — COMMENTED** [review]
I’d prefer for this to be merged only after or together with https://github.com/containers/image/pull/1035 to make sure the two are consistent.

_Right now_, the code LGTM, though the documentation does not match the code and needs updating.

---

The path lookup logic is complex enough that it probably makes sense to export `c/image/docker.configuredSignatureStorageBase` (with a better name) now, and use it here instead of maintaining a second copy of the code. (While overall we do want  readers/writers to use the c/image API to access signatures instead of raw access, this has clearly not stopped `podman sign`, and modifying `podman sign` to use a `c/image/docker` `ImageDestination` would now incompatibly introduce a requirement to be able to connect to the registry.)

**Comment by rhatdan** [comment]
@QiWang19 This needs a rebase.  Is this still being worked on?

**Comment by QiWang19** [comment]
@rhatdan still being worked on, waiting to get this commit from c/image https://github.com/containers/image/commit/1a0dda734e3bed323a9674e274b379cf841fe8b1

**Comment by rhatdan** [comment]
@QiWang19 you should have the containers/image PR in podman now.

**Comment by QiWang19** [comment]
@mtrmac @vrothberg PTAL

File: pkg/trust/trust.go
Comment (rhatdan) [inline pkg/trust/trust.go:89]: Should there be a rootless check here?

Comment (QiWang19) [reply pkg/trust/trust.go:89]: To keep consistent with part of the logic of c/image [newConfigWrapper ](https://github.com/containers/image/blob/f1a6f9b50d0286bf208935d4b9ab48117294ac31/pkg/sysregistriesv2/system_registries_v2.go#L351), if per-user dir exists, direclty use it, without checking whether its rootless or not

Comment (rhatdan) [reply pkg/trust/trust.go:89]: But what happens if the directory does not exists and you are rootless.  I would figure we still want to create it.

Comment (QiWang19) [reply pkg/trust/trust.go:89]: you mean need to create the $HOME/.config/containers/registries.conf.d if rootless? This  RegistriesDirPath only used by `podman image trust show` to see if the configuration for specific registry exist, so we don't change it if no  configuration exists. `podman image sign` will create the place for signature storage.

Comment (rhatdan) [reply pkg/trust/trust.go:89]: Ok I thought that was what was calling this.

**Comment by QiWang19** [comment]
@mtrmac PTAL

**Comment by QiWang19** [comment]
@containers/podman-maintainers PTAL

File: docs/source/markdown/podman-image-sign.1.md
Comment (TomSweeneyRedHat) [inline docs/source/markdown/podman-image-sign.1.md:12]:
```suggestion
derived from the registry configuration files in `$HOME/.config/containers/registries.d` if it exists,
```

File: docs/source/markdown/podman-image-sign.1.md
Comment (TomSweeneyRedHat) [inline docs/source/markdown/podman-image-sign.1.md:13]: looks like we don't list `containers-registries-d(5)` at the bottom of this man page and I think we should.

**Review by mtrmac — COMMENTED** [review]
These code changes LGTM.

To be consistent, though, I think this PR should also update `getPolicyShowOutput`  the other user of `trust.LoadAndMergeConfig`+`trust.HaveMatchRegistry` (and then delete them), so that that function also correctly shows the new built-in default paths.

A downside to that is that that function will now always show a sigstore, even for X-R-S-S registries (and there isn’t a way to tell whether a registry supports X-R-S-S without connecting to it), but that’s probably better than not showing a default sigstore path that would get used.

**Comment by QiWang19** [comment]
> These code changes LGTM.
> 
> To be consistent, though, I think this PR should also update `getPolicyShowOutput`, the other user of `trust.LoadAndMergeConfig`+`trust.HaveMatchRegistry` (and then delete them), so that that function also correctly shows the new built-in default paths.
> 

replace the above with `docker.SignatureStorageBaseURL` needs pass in an ImageReference, is there a good way to form an imageReference from the transports configurations in policy.json?

**Comment by mtrmac** [comment]
> > These code changes LGTM.
> > To be consistent, though, I think this PR should also update `getPolicyShowOutput`, the other user of `trust.LoadAndMergeConfig`+`trust.HaveMatchRegistry` (and then delete them), so that that function also correctly shows the new built-in default paths.
> 
> replace the above with `docker.SignatureStorageBaseURL` needs pass in an ImageReference, is there a good way to form an imageReference from the transports configurations in policy.json?

Ouch… no. There is one case that works ( a `policy.json` scope that includes a tag or digest); all the rest does not:
- For a host[:port]/…/repo scope, it would be possible to look up e.g. …/repo:latest, but that might not be correct if :latest has a different configuration section from the rest of the repo.
- For a host[:port], it is impossible to look up without inventing a repo+tag/digest, and that again could clash with a more specific configuration section.
- In the worst case, there will be new support for *.domain.com and the like in `policy.json`, so there could be multiple equally right answers if a.domain.com and b.domain.com are configured with different sigstores.

The last one is essentially unfixable given the current `trust.Policy` data structure and CLI output, so I suppose returning nothing if the existing code doesn’t find an exact match is as close as we can get.

**Review by mtrmac — APPROVED** [review]
LGTM.

**Comment by umohnani8** [comment]
Changes LGTM, @QiWang19 needs a rebase though

**Comment by QiWang19** [comment]
Needs lgtm label.

**Comment by umohnani8** [comment]
/lgtm
/hold

@mheon should we hold for the github-check failure?

**Comment by mheon** [comment]
We can't do anything about the RDO check, but it's not a blocking failure, we can merge without it.
/hold cancel

**Pull request merged**
